### PR TITLE
Entropy

### PR DESCRIPTION
**Pull request by amburger66**
- added "threshold" `interactive_query_policy`
- added "entropy" and "BALD" `interactive_score_function`s
- `MLPClassifierEnsemble` implements `logits()`
- a couple of scripts for possible experiments to run?

### CODE REVIEW

File: analysis/run_score_function_experiments.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+START_SEED=456
+NUM_SEEDS=10
+CYCLES=100
+FILE="analysis/submit.py"
+
+for SEED in $(seq $START_SEED $((NUM_SEEDS+START_SEED-1))); do
+    # section kid
```
Comment (tomsilver) [inline analysis/run_score_function_experiments.sh:9]: 🤣

File: analysis/run_score_function_experiments.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    python $FILE --env cover --approach interactive_learning --seed $SEED --excluded_predicates Covers --interactive_query_policy nonstrict_best_seen --interactive_score_function trivial --num_online_learning_cycles $CYCLES
+    # ours
+    python $FILE --env cover --approach interactive_learning --seed $SEED --excluded_predicates Covers --interactive_query_policy threshold --interactive_score_threshold 0.5 --interactive_score_function entropy --num_online_learning_cycles $CYCLES
+    # silent kid
```
Comment (tomsilver) [inline analysis/run_score_function_experiments.sh:13]: a truly silent kid would never query about anything, but this one may query one or a few times, right? can we change this to be truly silent, to make the eventual explanation of this baseline simpler?

also just to check -- this silent kid will still be learning things and changing over the course of online interactions, even though it's not querying, because we're adding all of the collected data into the set that's used for NSRT learning. right?

File: src/approaches/interactive_learning_approach.py [hunk 3]
```diff
@@ -264,6 +272,23 @@ def _query_policy(s: State) -> Optional[GroundAtomsHoldQuery]:
...
+
+        def _query_policy(s: State) -> Optional[GroundAtomsHoldQuery]:
+            ground_atoms = utils.all_possible_ground_atoms(
+                s, self._predicates_to_learn)
+            atoms_to_query = set()
+            for atom in ground_atoms:
+                score = self._score_atom_set({atom}, s)
+                if score > CFG.interactive_score_threshold:
+                    atoms_to_query.add(atom)
+                    self._best_score = score
```
Comment (tomsilver) [inline src/approaches/interactive_learning_approach.py:287]: is this needed?

File: src/torch_models.py [hunk 4]
```diff
@@ -453,15 +453,20 @@ def fit(self, X: Array, y: Array) -> None:
             member.fit(X, y)
 
     def classify(self, x: Array) -> bool:
-        member_vals = []
-        for member in self._members:
-            x_normalized = member.normalize(x)
-            member_vals.append(member(x_normalized).item())
-        avg = np.mean(member_vals)
+        avg = np.mean(self.logits(x))
```
Comment (tomsilver) [inline src/torch_models.py:456]: nice refactor!

File: src/torch_models.py [hunk 5]
```diff
@@ -453,15 +453,20 @@ def fit(self, X: Array, y: Array) -> None:
...
         assert classification in [True, False]
         return classification
 
+    def logits(self, x: Array) -> Array:
+        """Return logits calculated by each member."""
+        assert x.ndim == 1
+        logits = []
+        for member in self._members:
+            x_normalized = member.normalize(x)
+            logits.append(member(x_normalized).item())
```
Comment (tomsilver) [inline src/torch_models.py:467]: I'm getting turned around about whether the output of `member(*)` is a logit or just a probability. I think I stated the definition of logit wrong when we were talking about it earlier. A logit should be `log(p / (1- p))`, right? But I don't think that's what we want it to be here, nor do I think that's what it actually is -- I think this is really just the probability for class True. And for the calculation of entropy and the BALD objective, that's what we want it to be, too. If you agree, can we rename this function to `predict_proba`, in line with sklearn? Sorry for my confusion!

File: src/approaches/interactive_learning_approach.py [hunk 6]
```diff
@@ -303,3 +328,27 @@ def _score_atom_set_frequency(self, atom_set: Set[GroundAtom]) -> float:
...
+    def _score_atom_set_bald(self, atom_set: Set[GroundAtom],
+                             state: State) -> float:
+        """Score an atom set as the sum of the BALD objectives of each atom's
+        predicate."""
+        objective = 0.0
+        for atom in atom_set:
+            x = state.vec(atom.objects)
+            logits = self._pred_to_ensemble[atom.predicate.name].logits(x)
+            entropy = utils.entropy(np.mean(logits))
+            objective += entropy - np.mean(
```
Comment (tomsilver) [inline src/approaches/interactive_learning_approach.py:352]: looks good. btw, did you ever figure out what they were talking about w.r.t. computational advantages here?

**Review by tomsilver — COMMENTED** [review]
looking good! excited for these experiments! :)

File: src/approaches/interactive_learning_approach.py (hunk 3 above)
Comment (amburger66) [reply src/approaches/interactive_learning_approach.py:287]: oops nope

File: src/torch_models.py (hunk 5 above)
Comment (amburger66) [reply src/torch_models.py:467]: oh yes, that's my bad! i forgot that's the actual definition of logit. i will rename the function

File: src/approaches/interactive_learning_approach.py (hunk 6 above)
Comment (amburger66) [reply src/approaches/interactive_learning_approach.py:352]: hm so the original objective looks to maximize the decrease in expected posterior entropy (given the new data point x):
image
but the problems include: posteriors are high dimensional, calculating entropy is intractable, possibly infinite dimensional parameter spaces, expensive posterior updates.
however, that objective can be rewritten into the BALD (Bayesian Active Learning by Disagreement) objective (which one of these days i will verify for myself :P)
image
which calculates entropy in low-dim output space and only requires O(1) posterior updates

Comment (tomsilver) [reply src/approaches/interactive_learning_approach.py:352]: thank you! I'm not sure if I fully understand the details, but the high level makes sense -- let's talk through this in person. it looks good to me for now.

File: analysis/run_score_function_experiments.sh (hunk 2 above)
Comment (amburger66) [reply analysis/run_score_function_experiments.sh:13]: yep that makes sense, i'll make the silent kid truly silent :P

and yes, that's right

**Review by tomsilver — APPROVED** [review]
Great work!
